smarts/sstudio/genhistories.py

```diff
@@ -889,6 +932,41 @@ def constrain_angle(angle: float) -> float:
                 rows[j]["heading_rad"] = constrain_angle(rows[j]["heading_rad"])
                 yield rows[j]
 
+    def _encode_tl_state(self, waymo_state) -> SignalLightState:
+        if waymo_state == TrafficSignalLaneState.LANE_STATE_STOP:
+            return SignalLightState.STOP
+        if waymo_state == TrafficSignalLaneState.LANE_STATE_CAUTION:
+            return SignalLightState.CAUTION
+        if waymo_state == TrafficSignalLaneState.LANE_STATE_GO:
+            return SignalLightState.GO
+        if waymo_state == TrafficSignalLaneState.LANE_STATE_ARROW_STOP:
+            return SignalLightState.STOP | SignalLightState.ARROW
+        if waymo_state == TrafficSignalLaneState.LANE_STATE_ARROW_CAUTION:
+            return SignalLightState.CAUTION | SignalLightState.ARROW
+        if waymo_state == TrafficSignalLaneState.LANE_STATE_ARROW_GO:
+            return SignalLightState.GO | SignalLightState.ARROW
+        if waymo_state == TrafficSignalLaneState.LANE_STATE_FLASHING_STOP:
+            return SignalLightState.STOP | SignalLightState.FLASHING
+        if waymo_state == TrafficSignalLaneState.LANE_STATE_FLASHING_CAUTION:
+            return SignalLightState.CAUTION | SignalLightState.FLASHING
+        return SignalLightState.UNKNOWN
+
+    @property
+    def traffic_light_rows(self) -> Generator[Dict, None, None]:
+        scenario = self._get_scenario()
+        num_steps = len(scenario.timestamps_seconds)
+        for i in range(num_steps):
+            dynamic_states = scenario.dynamic_map_states[i]
+            for j, lane_state in enumerate(dynamic_states.lane_states):
```

Comment:
Looks like `j` is unused. Also we can probably pull the `sim_time` computation out of this loop.

Reply:
Done.